utils/filesystem/files_test.go

```diff
@@ -2030,3 +2030,261 @@ func TestConvertToOSFile(t *testing.T) {
+					globToRegex("*.o"),
+					globToRegex("*.jar"),
+				}
+				return fs.LsRecursiveWithExtensionPatterns(context.Background(), tree, exclusionPatterns...)
+			},
+			fields: fields{
+				vfs: vfs,
+			},
+			args: args{
+				ctx: context.Background(),
+				dir: rootDir,
+			},
+			wantFiles: []string{
+				filepath.Join(rootDir, "dir2", "dir3", "dir4", "test1.txt"),
+				filepath.Join(rootDir, "dir2", "dir3", "dir4", "test3.h"),
+				filepath.Join(rootDir, "dir2", "dir5", "test4.txt"),
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			name: "Test LsRecursiveWithExtensionPatternsAndLimits with enough max depth and enough max file count",
+			f: func(tree string) ([]string, error) {
+				limits := &Limits{MaxDepth: 4, MaxFileCount: 5, Recursive: true}
+				return fs.LsRecursiveWithExtensionPatternsAndLimits(context.Background(), tree, limits)
+			},
+			fields: fields{
+				vfs: vfs,
+			},
+			args: args{
+				ctx: context.Background(),
+				dir: rootDir,
+			},
+			wantFiles: []string{
+				filepath.Join(rootDir, "dir2", "dir3", "dir4", "test1.txt"),
+				filepath.Join(rootDir, "dir2", "dir3", "dir4", "test2.o"),
+				filepath.Join(rootDir, "dir2", "dir3", "dir4", "test3.h"),
+				filepath.Join(rootDir, "dir2", "dir5", "test4.txt"),
+				filepath.Join(rootDir, "dir2", "dir5", "test5.jar"),
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			name: "Test LsRecursiveWithExtensionPatternsAndLimits with enough max depth and enough max file count and exclusion patterns",
+			f: func(tree string) ([]string, error) {
+				limits := &Limits{MaxDepth: 4, MaxFileCount: 5, Recursive: true}
+				exclusionPatterns := []string{
+					globToRegex("*.o"),
+					globToRegex("*.h"),
+				}
+				return fs.LsRecursiveWithExtensionPatternsAndLimits(context.Background(), tree, limits, exclusionPatterns...)
+			},
+			fields: fields{
+				vfs: vfs,
+			},
+			args: args{
+				ctx: context.Background(),
+				dir: rootDir,
+			},
+			wantFiles: []string{
+				filepath.Join(rootDir, "dir2", "dir3", "dir4", "test1.txt"),
+				filepath.Join(rootDir, "dir2", "dir5", "test4.txt"),
+				filepath.Join(rootDir, "dir2", "dir5", "test5.jar"),
+			},
+			wantErr: assert.NoError,
+		},
+		{
+			name: "Test LsRecursiveWithExtensionPatternsAndLimits with NOT enough max depth and enough max file count",
+			f: func(tree string) ([]string, error) {
+				limits := &Limits{MaxDepth: 3, MaxFileCount: 5, Recursive: true}
+				return fs.LsRecursiveWithExtensionPatternsAndLimits(context.Background(), tree, limits)
+			},
+			fields: fields{
+				vfs: vfs,
+			},
+			args: args{
+				ctx: context.Background(),
+				dir: rootDir,
+			},
+			wantFiles: []string{
+				filepath.Join(rootDir, "dir2", "dir5", "test4.txt"),
+				filepath.Join(rootDir, "dir2", "dir5", "test5.jar"),
+			},
+			wantErr: func(t assert.TestingT, err error, msgAndArgs ...interface{}) bool {
+				if errors.Is(err, filepath.SkipDir) {
+					return true
+				}
+				return assert.NoError(t, err, msgAndArgs...)
+			},
```

Comment:
having functions returned like this makes it hard to follow when reading the test as you have to add conditional logic to parse the response